server/redeemer.go

```diff
@@ -0,0 +1,398 @@
+	r.subs.Range(
+		func(key, value interface{}) bool {
+			var maxFloatB []byte
+			if maxFloat != nil {
+				maxFloatB = maxFloat.Bytes()
+			}
+			value.(chan *net.MaxFloatUpdate) <- &net.MaxFloatUpdate{
+				Sender:   sender.Bytes(),
+				MaxFloat: maxFloatB,
+			}
+			return true
+		},
+	)
+}
+
+func (r *redeemer) MonitorMaxFloat(req *empty.Empty, stream net.TicketRedeemer_MonitorMaxFloatServer) error {
+	// The client address will serve as the ID for the stream
+	p, ok := peer.FromContext(stream.Context())
+	if !ok {
+		return status.Error(codes.Internal, "context is nil")
+	}
+
+	// Make a channel to receive max float updates
+	//  This check allows to overwrite the channel for testing purposes
+	var maxFloatUpdates chan *net.MaxFloatUpdate
+	maxFloatUpdatesI, ok := r.subs.Load(p.Addr.String())
+	if !ok {
+		maxFloatUpdates = make(chan *net.MaxFloatUpdate)
+		r.subs.Store(p.Addr.String(), maxFloatUpdates)
+		glog.Infof("new MonitorMaxFloat subscriber: %v", p.Addr.String())
+	} else {
+		maxFloatUpdates, ok = maxFloatUpdatesI.(chan *net.MaxFloatUpdate)
+		if !ok {
+			return status.Error(codes.Internal, "maxFloatUpdates is of the wrong type")
+		}
+	}
+
+	// Block so that the stream is over a long-lived connection
+	for {
+		select {
+		case maxFloatUpdate := <-maxFloatUpdates:
+			if err := stream.Send(maxFloatUpdate); err != nil {
+				if err == io.EOF {
+					r.subs.Delete(p.Addr.String())
+					return status.Error(codes.Internal, err.Error())
+				}
+				glog.Errorf("Unable to send maxFloat update to client=%v err=%v", p.Addr.String(), err)
+			}
+		case <-r.quit:
+			return nil
+		case <-stream.Context().Done():
+			return nil
+		}
+	}
+}
+
+func (r *redeemer) MaxFloat(ctx context.Context, req *net.MaxFloatRequest) (*net.MaxFloatUpdate, error) {
+	mf, err := r.sm.MaxFloat(ethcommon.BytesToAddress(req.Sender))
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Errorf("max float error: %v", err).Error())
+	}
+	return &net.MaxFloatUpdate{
+		Sender:   req.Sender,
+		MaxFloat: mf.Bytes(),
+	}, nil
+}
+
+func (r *redeemer) startCleanupLoop() {
+	ticker := time.NewTicker(cleanupLoopTime)
+	for {
+		select {
+		case <-ticker.C:
+			// clean up map entries that haven't been cleared since the last cleanup loop ran
+			r.liveSenders.Range(func(key, value interface{}) bool {
+				if value.(time.Time).Add(cleanupLoopTime).Before(time.Now()) {
+					r.liveSenders.Delete(key)
+				}
+				return true
+			})
+		case <-r.quit:
+			return
+		}
+	}
+}
+
+type redeemerClient struct {
+	rpc      net.TicketRedeemerClient
+	maxFloat map[ethcommon.Address]*big.Int
+	mu       sync.RWMutex
+	quit     chan struct{}
+	sm       pm.SenderManager
+	tm       pm.TimeManager
+}
+
+// NewRedeemerClient instantiates a new client for the ticket redemption service
+// The client implements the pm.SenderMonitor interface
+func NewRedeemerClient(uri *url.URL, sm pm.SenderManager, tm pm.TimeManager) (pm.SenderMonitor, *grpc.ClientConn, error) {
```

Comment:
I don't think we need to return the `SenderMonitor` interface here - the fact that `redeemerClient` implements the `SenderMonitor` interface can be implicit.

Reply:
Yeah sure we can export the struct instead

Reply:
fixed in 25b10e8d19440a55c9489a269e0e3c5e3858b352